packages/databus/src/index.ts

```diff
@@ -0,0 +1,135 @@
+/*-----------------------------------------------------------------------------
+| Copyright (c) Jupyter Development Team.
+| Distributed under the terms of the Modified BSD License.
+|----------------------------------------------------------------------------*/
+
+import { DisposableDelegate, IDisposable } from '@phosphor/disposable';
+
+import { ArrayExt } from '@phosphor/algorithm';
+
+import { ISignal, Signal } from '@phosphor/signaling';
+
+/**
+ * An interface for an abstract dataset, with a mimetype, data, and metadata.
+ *
+ * This is templated over the type of data `T` and the type of metadata `U`.
+ *
+ * #### Notes
+ * This interface is similar conceptually to that of nbformat.IMimeBundle,
+ * but with only a single mimetype and different structure. We expect there
+ * to be utilities to convert between this formats.
+ */
+export interface IDataset<T, U> {
+  /**
+   * The string mimetype for the dataset.
+   */
+  mimeType: string;
+
+  /***
+   * The data of the dataset.
+   *
+   * #### Notes
+   * The notion of data in the databus is completely abstract, and typed as an
+   * any. Publishers and consumers of a given mimetype will need to agree
+   * what the actual data is and cast it appropriately before use.
+   */
+  data: T;
+
+  /**
+   * The metadata associate with the dataset.
+   *
+   * #### Notes
+   * Like the data itself, the databus notion of metadata is entirely abstract.
+   * We are currently exploring different metadata schemes, so this may change
+   * in the future.
+   */
+  metadata: U;
+}
+
+/**
+ * A databus object, for managing datasets in JupyterLab.
+ */
+export class DataBus {
+  /**
+   * Construct a new databus.
+   */
+  constructor() {}
+
+  /**
+   * Publish a dataset to the databus.
+   *
+   * @param dataset - the `IDataset` to publish to the databus.
+   *
+   * @returns A disposable which will remove the dataset.
+   *
+   * @throws An error if the given dataset is already published.
+   */
+  publish(dataset: IDataset<any, any>): IDisposable {
+    if (this._datasets.find(i => i === dataset) === undefined) {
+      throw new Error(`Dataset already published`);
+    }
+
+    this._datasets.push(dataset);
+
+    this._datasetsChanged.emit({ dataset, type: 'added' });
+
+    return new DisposableDelegate(() => {
+      ArrayExt.removeFirstOf(this._datasets, dataset);
+      this._datasetsChanged.emit({ dataset, type: 'removed' });
+    });
+  }
+
+  /**
+   * Filter the published datasets by a mime type.
+   *
+   * @param mimeType - The `mimeType` to filter on.
+   *
+   * @returns An array of matching `IDataset` objects.
+   */
+  filter(mimeType: string): Array<IDataset<any, any>> {
+    return this._datasets.filter(dataset => {
+      if ((dataset as IDataset<any, any>).mimeType === mimeType) {
+        return dataset;
+      }
+    });
+  }
+
+  /**
+   * Return a list of all published datasets.
+   */
+  get datasets(): Array<IDataset<any, any>> {
```

Comment:
Sometimes with a datastructure like this we expose an iterator, somtimes an array. I don't think I have a *super* strong gut feeling for when we should choose one over the other, but an argument for an iterator could be

a. when the number of datasets could be dynamic or unknown, and
b. when we want to emphasize that the ordering is meaningless and should not be relied upon.

What would you say to making this an iterator instead?

Reply:
Another option would be to make this an [iterable](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Guide/Iterators_and_Generators#User-defined_iterables), which would allow users to loop through it or turn it into an array if they would like (with `[...databus]`).

Reply:
Don't feel strongly about using Array, but I do think it is important to emphasize the the order should not be relied upon. I can look into the iterator and iterable.